src/product_config_utils.rs

```diff
@@ -0,0 +1,1034 @@
+/// - Configuration files (files)
+///
+/// Returned empty Maps will be ignored. Check out `to_hash_map(...)` in
+/// https://github.com/stackabletech/product-config/blob/main/src/ser.rs
+/// if you do not need to differentiate between the options and want to serialize all properties
+/// in a certain config [Config] to one option.
+///
+pub trait Configuration {
+    type Configurable;
+
+    // TODO: Not sure if I need the role_name here
+    // TODO: We need to pass in the existing config from parents to run validation checks and we should probably also pass in a "final" parameter or have another "finalize" method callback
+    //  one for each role group, one for each role and one for all of it...
+    fn compute_env(
+        &self,
+        resource: &Self::Configurable,
+        role_name: &str,
+    ) -> Result<BTreeMap<String, String>, ConfigError>;
+
+    fn compute_cli(
+        &self,
+        resource: &Self::Configurable,
+        role_name: &str,
+    ) -> Result<BTreeMap<String, String>, ConfigError>;
+
+    fn compute_files(
+        &self,
+        resource: &Self::Configurable,
+        role_name: &str,
+        file: &str,
+    ) -> Result<BTreeMap<String, String>, ConfigError>;
+}
+
+// This deep map causes problems with clippy and fmt.
+pub type RoleConfigByPropertyKind =
+    HashMap<String, HashMap<String, HashMap<PropertyNameKind, BTreeMap<String, String>>>>;
+///
+/// Given the configuration parameters of all `roles` partition them by `PropertyNameKind` and
+/// merge them with the role groups configuration parameters.
+///
+/// The output is a map keyed by the role names. The value is also a map keyed by role group names and
+/// the values are the merged configuration properties "bucketed" by `PropertyNameKind`.
+/// # Arguments
+/// - `resource`         - Not used directly. It's passed on to the `Configuration::compute_*` calls.
+/// - `role_information` - A map keyed by role names. The value is a vector of `PropertyNameKind`
+/// - `roles`            - A map keyed by role names.
+///
+pub fn transform_all_roles_to_config<T>(
+    resource: &T::Configurable,
+    role_information: HashMap<String, Vec<PropertyNameKind>>,
+    roles: HashMap<String, Role<T>>,
+) -> RoleConfigByPropertyKind
+where
+    T: Configuration,
+{
+    let mut result = HashMap::new();
+
+    for (role_name, role) in roles {
+        let property_name_kinds = match role_information.get(&role_name) {
+            Some(kind) => kind,
+            None => {
+                error!("The role [{}] was not specified in [{:?}]. This is a programming error please report a ticket. Will skip for now.", role_name, role_information.keys());
+                continue;
+            }
+        };
+
+        let role_properties =
+            transform_role_to_config(resource, &role_name, &role, property_name_kinds);
+        result.insert(role_name, role_properties);
+    }
+
+    result
+}
+
+// TODO: boolean flags suck, move ignore_warn to be a flag
```

Comment:
I don't think boolean arguments are outright banned, but if are commonly passing around booleans with the same semantics and/or if you are passing multiple booleans in the same function, using enums will definitely help. (I also will sometimes name the value in the caller, for example have `let ignore_warn = true; process_validation_results(_, ignore_warn, _)`). Here it also looks like `ignore_warn` and `ignore_err` potentially encode some redundant state that could be invalid (would it be valid for a caller to call this with `false, true`?), in which case it could definitely make sense to encode this differently (maybe in a tri-state enum or in a custom enum wrapped in `Option`).

Reply:
Yeah i agree, for now the (false, true) state seems a litte odd. I think we can keep it for now (as we just set it globally in the operators. But we should move to a more granular method to distinguish ignore warn / errors for each property like lars proposed:
`pub enum Property {
    Simple(String),
    Complex {
        ignore_warning: bool,
        ignore_error: bool,
        value: String,
    },
}`